### PR TITLE
Operator status revisions

### PR DESCRIPTION
**Pull request by benjaminapetersen**
Extracted a bunch of operator status updates out into separate named functions.  Was a bit pedantic about naming, but hopefully it helps understand the flow and see if we can tease out any bugs.

@jhadvig

### CODE REVIEW

File: pkg/console/operator/operator.go [hunk 1]
```diff
@@ -169,22 +181,22 @@ func (c *consoleOperator) handleSync(operatorConfig *operatorsv1.Console, consol
 		return fmt.Errorf("unknown state: %v", operatorConfig.Spec.ManagementState)
 	}
 
-	// do we need the if(configChanged) update bits?
-	operatorConfigOut, consoleConfigOut, configChanged, err := sync_v400(c, operatorConfig, consoleConfig)
+	_, _, _, err := sync_v400(c, operatorConfig, consoleConfig)
```
Comment (benjaminapetersen) [inline pkg/console/operator/operator.go]: I didn't notice this change, I need to look into it a bit.

File: pkg/console/operator/status.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
+package operator
```
Comment (benjaminapetersen) [inline pkg/console/operator/status.go]: These are pretty verbose / repetitive.

File: pkg/console/operator/sync_v400.go [hunk 3]
```diff
@@ -49,46 +50,72 @@ func sync_v400(co *consoleOperator, operatorConfig *operatorv1.Console, consoleC
...
-	_, depChanged, depErr := SyncDeployment(co, recorder, operatorConfig, cm, serviceCAConfigMap, sec)
+	actualDeployment, depChanged, depErr := SyncDeployment(co, recorder, operatorConfig, cm, serviceCAConfigMap, sec)
 	if depErr != nil {
+		co.operatorStatusResourceSyncFailure(operatorConfig, fmt.Sprintf("%q: %v\n", "route", depErr))
 		return operatorConfig, consoleConfig, toUpdate, depErr
 	}
+
 	toUpdate = toUpdate || depChanged
 
+	// at this point, we should not be failing anymore
```
Comment (benjaminapetersen) [inline pkg/console/operator/sync_v400.go]: Does this now make sense? Thinking through it.

**Comment by benjaminapetersen** [comment]
Note that all of these `ClusterOperator` PRs will cause the `Console` operator to appear on the `Cluster Settings` page, under `Cluster Operators`:

screen shot 2019-02-14 at 3 38 53 pm

**Comment by benjaminapetersen** [comment]
A test:
`oc delete deployment console -n openshift-console`
`oc get events -n openshift-console-operator` 

shows:
```
LAST SEEN   TYPE      REASON                  KIND         MESSAGE
9m11s       Normal    Scheduled               Pod          Successfully assigned openshift-console-operator/console-operator-779cc894db-8b8vz to ip-10-0-133-68.us-west-1.compute.internal
9m3s        Normal    Pulling                 Pod          pulling image "quay.io/benjaminapetersen/console-operator:latest"
8m58s       Normal    Pulled                  Pod          Successfully pulled image "quay.io/benjaminapetersen/console-operator:latest"
8m58s       Normal    Created                 Pod          Created container
8m58s       Normal    Started                 Pod          Started container
9m11s       Normal    Killing                 Pod          Killing container with id cri-o://console-operator:Need to kill Pod
9m11s       Normal    SuccessfulCreate        ReplicaSet   Created pod: console-operator-779cc894db-8b8vz
8m39s       Normal    LeaderElection          ConfigMap    d4803d32-3097-11e9-a364-0a580a81020b became leader
8m39s       Warning   StatusNotFound          Deployment   Unable to determine current operator status for console
8m39s       Normal    OperatorStatusChanged   Deployment   Status for operator console changed: Failing set to False (""),Available set to True (""),Progressing set to True (""),status.relatedObjects changed from [] to [{"operator.openshift.io" "consoles" "" "console"} {"config.openshift.io" "consoles" "" "console"} {"oauth.openshift.io" "oauthclients" "" "console"} {"" "namespaces" "" "openshift-console-operator"} {"" "namespaces" "" "openshift-console"}]
8m39s       Normal    OperatorStatusChanged   Deployment   Status for operator console changed: Progressing changed from True to False (""),status.relatedObjects changed from [] to [{"operator.openshift.io" "consoles" "" "console"} {"config.openshift.io" "consoles" "" "console"} {"oauth.openshift.io" "oauthclients" "" "console"} {"" "namespaces" "" "openshift-console-operator"} {"" "namespaces" "" "openshift-console"}]
34s         Normal    DeploymentCreated       Deployment   Created Deploymentapps./console -n openshift-console because it was missing
34s         Normal    OperatorStatusChanged   Deployment   Status for operator console changed: Progressing changed from False to True ("")
34s         Normal    OperatorStatusChanged   Deployment   Status for operator console changed: Progressing changed from True to False ("")
34s         Normal    OperatorStatusChanged   Deployment   Status for operator console changed: Available changed from True to False ("NoDeploymentPodsAvailableOnAnyNode.")
32s         Normal    OperatorStatusChanged   Deployment   Status for operator console changed: Available changed from False to True ("")
```
Which is interesting, and 
` oc describe clusteroperator console`
Will result in:
```
Name:         console
Namespace:
Labels:       <none>
Annotations:  <none>
API Version:  config.openshift.io/v1
Kind:         ClusterOperator
Metadata:
  Creation Timestamp:  2019-02-14T20:34:04Z
  Generation:          1
  Resource Version:    113010
  Self Link:           /apis/config.openshift.io/v1/clusteroperators/console
  UID:                 de84e7e6-3097-11e9-968a-02a640b76a0a
Spec:
Status:
  Conditions:
    Last Transition Time:  2019-02-14T20:42:24Z
    Status:                True
    Type:                  Available
    Last Transition Time:  2019-02-14T20:42:09Z
    Status:                False
    Type:                  Progressing
    Last Transition Time:  2019-02-14T20:34:04Z
    Status:                False
    Type:                  Failing
  Extension:               <nil>
  Related Objects:
    Group:     operator.openshift.io
    Name:      console
    Resource:  consoles
    Group:     config.openshift.io
    Name:      console
    Resource:  consoles
    Group:     oauth.openshift.io
    Name:      console
    Resource:  oauthclients
    Group:
    Name:      openshift-console-operator
    Resource:  namespaces
    Group:
    Name:      openshift-console
    Resource:  namespaces
  Versions:    <nil>
Events:        <none>
```
Which seems promising.

**Comment by benjaminapetersen** [comment]
Fail message:

```
level=warning msg="Found override for ReleaseImage. Please be warned, this is not advised"
level=info msg="Consuming \"Install Config\" from target directory"
level=info msg="Creating cluster..."
level=info msg="Waiting up to 30m0s for the Kubernetes API..."
level=info msg="API v1.12.4+bb5f71c up"
level=info msg="Waiting up to 30m0s for the bootstrap-complete event..."
level=warning msg="RetryWatcher - getting event failed! Re-creating the watcher. Last RV: 4768"
level=info msg="Destroying the bootstrap resources..."
level=info msg="Waiting up to 30m0s for the cluster to initialize..."
level=fatal msg="failed to initialize the cluster: Cluster operator console has not yet reported success"
```

File: pkg/console/operator/sync_v400.go (hunk 3 above)
Comment (jhadvig) [reply pkg/console/operator/sync_v400.go]: it does, I'm counting on it as well

File: pkg/console/operator/sync_v400.go [hunk 4]
```diff
@@ -49,46 +50,72 @@ func sync_v400(co *consoleOperator, operatorConfig *operatorv1.Console, consoleC
...
+	co.operatorStatusResourceSyncSuccess(operatorConfig)
+
+	// but we may be in a transitional state, if any of the above resources changed
+	if toUpdate {
+		co.operatorStatusProgressing(operatorConfig)
```
Comment (jhadvig) [inline pkg/console/operator/sync_v400.go]: was debugging a bit more and figured our that even if none of the resources are updated, here the `toUpdate` will be set to `true`, which shouldn't happen if nothing gets updated, right ?

Comment (benjaminapetersen) [reply pkg/console/operator/sync_v400.go]: Nope, if nothing is updated, we want it `false`.  So that is definitely something worth tracking down.

**Comment by benjaminapetersen** [comment]
Just pasting again....

So this error:

`level=fatal msg="failed to initialize the cluster: Cluster operator console has not yet reported success"`

https://github.com/openshift/cluster-version-operator/blob/9335e43c21efc9df7eed48f05585997087fc04db/pkg/cvo/internal/operatorstatus.go#L126

Means we need `Available: true, Progressing: false, Failing: false` in order for CVO to be happy & tell the installer its happy.

**Comment by benjaminapetersen** [comment]
/retest

the brand fix PR went in.

**Comment by benjaminapetersen** [comment]
Prefer #142 for tidiness (squashed verison of this)

**Comment by benjaminapetersen** [comment]
Closing, we have a newer branch that has more updates that we need.

**Pull request closed without merging**